src/compiler/factory.ts

```diff
@@ -3601,6 +3595,34 @@ namespace ts {
         return expression;
     }
 
+    export function parenthesizeElementTypeMember(member: TypeNode) {
+        switch (member.kind) {
+            case SyntaxKind.UnionType:
+            case SyntaxKind.IntersectionType:
+            case SyntaxKind.FunctionType:
+            case SyntaxKind.ConstructorType:
+                return createParenthesizedType(member);
+        }
+        return member;
+    }
+
+    export function parenthesizeElementTypeMembers(members: TypeNode[]) {
+        // TODO: does this lose `originalNode` ptr?
+        return createNodeArray(members.map(parenthesizeElementTypeMember));
+    }
+
+    export function parenthesizeTypeParameters(typeParameters: TypeNode[]) {
+        if (typeParameters && typeParameters.length > 0) {
+            const nodeArray = createNodeArray(typeParameters);
+            const firstEntry = nodeArray[0];

```

Comment:
Create a fresh node array, `createNodeArray` actually just turns `typeParameters` into a node array, so you end up mutating an existing array.